File: magic_square_method/magic_square.c
#include <stdio.h>
#include <stdlib.h>

// creates a 2D square array of 0s with given number of rows
int **createGrid(int n) {
    int i, j;

    // 2D array of ints
    int **matrix;

    matrix = malloc(sizeof(int *) * n);

    for (i = 0; i < n; i++) {
        matrix[i] = malloc(sizeof(int) * n);
    }

    // fill the matrix with 0s
    for (i = 0; i < n; i++) {
        for (j = 0; j < n; j++) {
            matrix[i][j] = 0;
        }
    }

    return matrix;
}

// creates a magic square array of given n
int **magicSquare(int **arr, int num) {
    int i, j;
    int startRow, startCol;

    startRow = 0;
    startCol = num / 2;

    arr[startRow][startCol] = 1;    // adding the 1

    for (i = 2; i < num * num + 1; i++) {
        int newRow = startRow - 1;  // one step up
        int newCol = startCol - 1;  // one step left

        // if the row number becomes negative then make it last row
        if (newRow < 0) {
            newRow += num;  
        }

        // if the column number becomes negative then make it last column
        if (newCol < 0) {
            newCol += num;
        }

        // if the cell is equipped already
        if (arr[newRow][newCol] != 0) {
            newRow = startRow + 1;  // one step down
            newCol = startCol;  // same column
        }

        // filling the array
        arr[newRow][newCol] = i;

        // updating startRow, startCol variables
        startRow = newRow;
        startCol = newCol;
    }

    return arr;
}

// prints the 2D array
void printArray(int **arr, int n) {
    int i, j;
    for (i = 0; i < n; i++) {
        for (j = 0; j < n; j++) {
            printf("%d\t", arr[i][j]);
        }
        printf("\n");
    }
}

void main() {
    int n;
    int i, j;
    int **arr;

    printf("Enter the value of n: ");
    scanf("%d", &n);

    // check if the n's value even
    if (n % 2 == 0) {
        printf("n's value should not be even!!!\n");
        return;
    }

    arr = createGrid(n);    // creating a 2D array of 0s
    arr = magicSquare(arr, n);  // magic square 

    printf("The magic-square for n = %d is:\n", n);
    printArray(arr, n);

    // free the memory
    for (i = 0; i < n; i++) {
        free(arr[i]);
    }
    free(arr);
}

File: magic_square_method/magic_square.py
from tabulate import tabulate
import time


list_of_list_of_int = list[list[int, ...], ...]

def create_grid(n: int = 3) -> list_of_list_of_int:
    """
    Creates a 2D list of 0s
    """

    assert n % 2 != 0, "n should not be an even number!!!"
    
    grid: list_of_list_of_int = [[0 for _ in range(n)] for _ in range(n)]
    return grid


def magic_square(arr: list_of_list_of_int) -> list_of_list_of_int:
    """
    Fill the grid using magic-square method
    """

    # size of the array
    n: int = len(arr)

    # for starting with 1 at the middle column and first row
    start_row: int = 0
    start_col: int = n // 2

    # starting with 1
    arr[start_row][start_col] = 1

    # fill up the others
    for i in range(2, n ** 2 + 1):
        new_row: int = start_row - 1 # move one up
        new_col: int = start_col - 1 # move one left

        if arr[new_row][new_col] == 0:
            if new_row < 0: # if new row becomes negative
                new_row = n + new_row # new row becomes the last row

            if new_col < 0: # if new column becomes negative
                new_col = n + new_col # new column becomes the last column
        else:
            new_row = start_row + 1 # move one down
            new_col = start_col # same column

        # updating the array values 
        arr[new_row][new_col] = i

        # updating start_row and start_column
        start_row = new_row
        start_col = new_col

    return arr
    
if __name__ == "__main__":
    start_time = time.perf_counter()
    
    n: int = 5
    arr: list_of_list_of_int = create_grid(n)
    arr: list_of_list_of_int = magic_square(arr)
    
    print(f"Execution Time: {(time.perf_counter() - start_time)} seconds\n")
    print(f"The magic-square table for {n = } is: ")
    print(tabulate(arr, tablefmt="grid"))
